### PR TITLE
Fix Quarkus ClassLoader's jar ProtectionDomain handling

### LINKED ISSUE #11549: Quarkus, Jackson and Kotlin 1.4
**Describe the bug**
* Upgraded to Kotlin 1.4
* Use Jackson and Jackson-module-kotlin
* Running with @QuarkusTest fails
* Same test without @QuarkusTest is ok

Se repository: [https://github.com/kny78/quarkus-kotlin14-jackson-fail](https://github.com/kny78/quarkus-kotlin14-jackson-fail)
See failure output: [https://github.com/kny78/quarkus-kotlin14-jackson-fail/tree/master/failure-logs](https://github.com/kny78/quarkus-kotlin14-jackson-fail/tree/master/failure-logs)

**Expected behavior**
Both tests should run ok:
* im.kny.DeSerializeQuarkusTest (Fails)
* im.kny.DeSerializeNonQuarkusTest (Ok)

**Actual behavior**
The test im.kny.DeSerializeQuarkusTest fail.s

**To Reproduce**
1. git clone git@github.com:kny78/quarkus-kotlin14-jackson-fail.git
2. mvnw install
3. 

```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:

        Linux kjetil 5.7.15-200.fc32.x86_64 #1 SMP Tue Aug 11 16:36:14 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: 

        openjdk version "11.0.8" 2020-07-14
        OpenJDK Runtime Environment 18.9 (build 11.0.8+10)
        OpenJDK 64-Bit Server VM 18.9 (build 11.0.8+10, mixed mode, sharing)

 - GraalVM version (if different from Java): N/A
 - Quarkus version or git rev: 1.7.0
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 

        Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
        Maven home: /home/kny/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
        Java version: 11.0.8, vendor: Red Hat, Inc., runtime: /usr/lib/jvm/java-11-openjdk-11.0.8.10-2.fc32.x86_64
        Default locale: en_GB, platform encoding: UTF-8
        OS name: "linux", version: "5.7.15-200.fc32.x86_64", arch: "amd64", family: "unix"


**Additional context**
(Add any other context about the problem here.)

### PR DESCRIPTION
**Pull request by geoand**
Bring it inline with what `URLClassLoader` does

Fixes: #11549

### CODE REVIEW

**Comment by geoand** [comment]
@stuartwdouglas I assume that the current code was just the result of a bug? Or did you have something specific in mind?

**Comment by geoand** [comment]
The CI **is** due to this PR - looking at it now

**Comment by geoand** [comment]
CI issue should now be fixed - Pinging @jaikiran as well since I made a change to his code

**Comment by jaikiran** [comment]
I'm a bit surprised that it's failing in JRE, because from what I can tell, I don't think the current code in `JarClassPathElement` is doing anything incorrect. Let me see if I can reproduce this standalone outside of Quarkus.

**Comment by geoand** [comment]
@jaikiran https://github.com/quarkusio/quarkus/issues/11549#issuecomment-683918475 (which I was able to reproduce) was the reason I made this change

**Comment by jaikiran** [comment]
I could reproduce this outside of Quarkus in a simple Java class. So I asked for inputs from the JDK team[1]. The initial assessment is that, the construct used by Kotlin(?) and that other library in the linked issue:
```
new JarInputStream(kotlinUnit.getProtectionDomain().getCodeSource().getLocation().openStream())
```
isn't expected to work when the URL in question is a `jar:file` URL.

So I went ahead and tried out simple test to see how URLClassLoader behaves when it comes to `CodeSource` location and what URL it returns for classes loaded out of jar files in the classpath. My tests show that it matches with your changes and it returns a plain `file:` URL (example `file:/tmp/some-jar.jar`).

So this change looks fine to me.

[1] https://mail.openjdk.java.net/pipermail/net-dev/2020-September/014389.html

**Comment by geoand** [comment]
Awesome, thanks for the great analysis @jaikiran!

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java [hunk 1]
```diff
@@ -217,7 +218,8 @@ void registerExchangeAttributeBuilders(final BuildProducer<ServiceProviderBuildI
             logger.debug("Skipping registration of service providers for " + ExchangeAttributeBuilder.class);
             return;
         }
-        try (final FileSystem jarFileSystem = ZipUtils.newFileSystem(codeSource.getLocation().toURI(),
+        try (final FileSystem jarFileSystem = ZipUtils.newFileSystem(
+                new URI("jar", codeSource.getLocation().toURI().toString(), null),
```
Comment (jaikiran) [inline extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java:222]: These `URI`/`URL` APIs are sometimes a bit too complex and I usually have to go back read their docs multiple times and even try out a few things to be sure that whatever construct is being used/changed will work fine without issues. So, I'm unsure if this change will cause any issues. However, given that this specific area of code has relevant tests[1] that cover Java 8, 11, 14 and even native-image, I think we can go ahead with this change if the CI shows all green.

[1] https://github.com/quarkusio/quarkus/pull/9474

**Comment by geoand** [comment]
CI failure seems totally unrelated, but since this change could potentially have a wide impact, I will restart CI since I want to be totally sure

**Review by jaikiran — APPROVED** [review]

File: extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/VertxHttpProcessor.java:222]: Agreed :)

**Comment by gsmet** [comment]
@stuartwdouglas I let you check this one? I will backport it tomorrow morning.

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**